crypto/x509/x509_cmp.c

```diff
@@ -440,19 +440,23 @@ int X509_CRL_check_suiteb(X509_CRL *crl, EVP_PKEY *pk, unsigned long flags)
 }
 
 #endif
+
+static X509 *x509upref(X509 *x)
+{
+    if (!X509_up_ref(x))
+        return NULL;
+
+    return x;
+}
+
 /*
  * Not strictly speaking an "up_ref" as a STACK doesn't have a reference
  * count but it has the same effect by duping the STACK and upping the ref of
  * each X509 structure.
  */
 STACK_OF(X509) *X509_chain_up_ref(STACK_OF(X509) *chain)
 {
-    STACK_OF(X509) *ret;
-    int i;
-    ret = sk_X509_dup(chain);
-    for (i = 0; i < sk_X509_num(ret); i++) {
-        X509 *x = sk_X509_value(ret, i);
-        X509_up_ref(x);
-    }
-    return ret;
+    /* We upref each element rather than actually copying */
+    return sk_X509_deep_copy(chain, (X509 *(*)(const X509 *))x509upref,
```

Comment:
I'm not a big fan of pretending that we're not modifying the original, and have to wonder how the compiler might optimize it.